### PR TITLE
Support `gitea_oauth2_app` `confidential_client` flag

### PR DESCRIPTION
**Pull request by deblasis**
Hi there, thank you for this provider!

I found out that there's a missing field in `gitea_oauth2_app`.

https://try.gitea.io/api/swagger#/user/userCreateOAuth2Application 
image

Currently it's implicitly passed as `false` therefore creating apps that require PKCE. This PR simply allows the consumer to configure the value.

From the swagger file above it looks like the default should be `true`, not a big deal but I think it's worth mentioning it.
If we set the default to `true` it would be a breaking change since it could break some consumers that might expect it to false since now it's ignored.

Reference: https://docs.gitea.com/next/development/oauth2-provider#confidential-client

### CODE REVIEW

**Comment by Lerentis** [comment]
Hi @deblasis ,
thanks for raising and picking this up already. you probably want to raise this to https://gitea.com/gitea/go-sdk as well as i am just using this SDK under the hood and you changed it there 😃

File: gitea/resource_gitea_oauth_app.go [hunk 1]
```diff
@@ -8,10 +8,11 @@ import (
 )
 
 const (
-	oauth2KeyName         string = "name"
-	oauth2KeyRedirectURIs string = "redirect_uris"
-	oauth2KeyClientId     string = "client_id"
-	oauth2KeyClientSecret string = "client_secret"
+	oauth2KeyName               string = "name"
+	oauth2KeyConfidentialClient string = "confidential_client"
```
Comment (Lerentis) [inline gitea/resource_gitea_oauth_app.go:12]: You probably want to add this to at least the [create options](https://github.com/Lerentis/terraform-provider-gitea/blob/main/gitea/resource_gitea_oauth_app.go#L92), else this change does not have any effect

**Review by Lerentis — CHANGES_REQUESTED** [review]

File: gitea/resource_gitea_oauth_app.go (hunk 1 above)
Comment (deblasis) [reply gitea/resource_gitea_oauth_app.go:12]: 🤦‍♂️ sorry about that, I tested the underlying vendored sdk change but not the terraform provider itself. I looked into that quickly and it sounded laborious to create a private registry, unless there's an easier way to use a custom provider that I am missing. 

TL;DR; I was lazy :)

I will rectify

Comment (deblasis) [reply gitea/resource_gitea_oauth_app.go:12]: Maybe addressed in [c707cba](https://github.com/Lerentis/terraform-provider-gitea/pull/46/commits/c707cba22ef1b80e4f6783bbcbd1d6bbea1bae92)

PTAL 🙏

File: gitea/resource_gitea_oauth_app.go [hunk 2]
```diff
@@ -176,9 +190,10 @@ func setOAuth2ResourceData(app *gitea.Oauth2, d *schema.ResourceData) (err error
 	d.SetId(app.ClientID)
 
 	for k, v := range map[string]interface{}{
-		oauth2KeyName:         app.Name,
-		oauth2KeyRedirectURIs: schema.NewSet(schema.HashString, CollapseStringList(app.RedirectURIs)),
-		oauth2KeyClientId:     app.ClientID,
+		oauth2KeyName:               app.Name,
+		oauth2KeyConfidentialClient: app.ConfidentialClient,
```
Comment (Lerentis) [inline gitea/resource_gitea_oauth_app.go:194]: this leads to a nil pointer dereference if the flag is not set

**Review by Lerentis — CHANGES_REQUESTED** [review]
the create looks good now, but while testing in encountered a nil pointer dereference using the following test config: 

```terraform
resource "gitea_oauth2_app" "test_oauth_app" {
  name = "test"
  redirect_uris = ["https://localhost"]
  confidential_client = true #<- changed this after create
}

output "oauth" {
  value = resource.gitea_oauth2_app.test_oauth_app.client_secret
  sensitive = true
}
``` 

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x9260d9]

goroutine 42 [running]:
git.uploadfilter24.eu/terraform-provider-gitea/gitea.setOAuth2ResourceData(0x0, 0xc0004b2150?)
        /home/lerentis/go/src/git.uploadfilter24.eu/lerentis/terraform-provider-gitea/gitea/resource_gitea_oauth_app.go:190 +0x39
git.uploadfilter24.eu/terraform-provider-gitea/gitea.resourceOauth2AppUpcreate(0xc000481700, {0xcb8580?, 0xc000620210})
        /home/lerentis/go/src/git.uploadfilter24.eu/lerentis/terraform-provider-gitea/gitea/resource_gitea_oauth_app.go:129 +0x45f
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).update(0xddaf40?, {0xddaf40?, 0xc000552210?}, 0xd?, {0xcb8580?, 0xc000620210?})
        /home/lerentis/go/src/git.uploadfilter24.eu/lerentis/terraform-provider-gitea/vendor/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema/resource.go:755 +0x178
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc000200b60, {0xddaf40, 0xc000552210}, 0xc0004ba5b0, 0xc000481580, {0xcb8580, 0xc000620210})
        /home/lerentis/go/src/git.uploadfilter24.eu/lerentis/terraform-provider-gitea/vendor/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema/resource.go:874 +0x845
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc000413d70, {0xddaf40?, 0xc0005520f0?}, 0xc000180d20)
        /home/lerentis/go/src/git.uploadfilter24.eu/lerentis/terraform-provider-gitea/vendor/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema/grpc_provider.go:1024 +0xe8d
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0xc0002cebe0, {0xddaf40?, 0xc0005336e0?}, 0xc00016e690)
        /home/lerentis/go/src/git.uploadfilter24.eu/lerentis/terraform-provider-gitea/vendor/github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server/server.go:821 +0x574
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0xc8be20?, 0xc0002cebe0}, {0xddaf40, 0xc0005336e0}, 0xc00016e620, 0x0)
        /home/lerentis/go/src/git.uploadfilter24.eu/lerentis/terraform-provider-gitea/vendor/github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:422 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0002c61e0, {0xddead8, 0xc00048a1a0}, 0xc000548240, 0xc00041d020, 0x12bc460, 0x0)
        /home/lerentis/go/src/git.uploadfilter24.eu/lerentis/terraform-provider-gitea/vendor/google.golang.org/grpc/server.go:1337 +0xdf3
google.golang.org/grpc.(*Server).handleStream(0xc0002c61e0, {0xddead8, 0xc00048a1a0}, 0xc000548240, 0x0)
        /home/lerentis/go/src/git.uploadfilter24.eu/lerentis/terraform-provider-gitea/vendor/google.golang.org/grpc/server.go:1714 +0xa36
google.golang.org/grpc.(*Server).serveStreams.func1.1()
        /home/lerentis/go/src/git.uploadfilter24.eu/lerentis/terraform-provider-gitea/vendor/google.golang.org/grpc/server.go:959 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
        /home/lerentis/go/src/git.uploadfilter24.eu/lerentis/terraform-provider-gitea/vendor/google.golang.org/grpc/server.go:957 +0x18c
```

i marked the point with an idea why it occurred

**Comment by deblasis** [comment]
Hi @Lerentis, in the end... I learnt how to debug terraform providers 😀

I identified the bug, unrelated to my changes but regardless, here's the fix: https://github.com/Lerentis/terraform-provider-gitea/pull/46/commits/2d7147442bbfd0c613bfe2c6d42c1e8194c3b494

The cause is just incorrect variable scope in the "update path".

Demo:
https://github.com/Lerentis/terraform-provider-gitea/assets/29378614/8bad7cff-0964-4d99-8997-53bbf58fd93d

**Comment by Lerentis** [comment]
woah kudos for finding this and fixing this as well 💯

**Review by Lerentis — APPROVED** [review]
LGTM
